### PR TITLE
OCPI: do not return locations with no charging stations.

### PR DESCRIPTION
**Pull request by jerome-benoit**
Signed-off-by: Jérôme Benoit <jerome.benoit@sap.com>

### CODE REVIEW

File: src/migration/tasks/AddSiteAreaLimitToConsumptionsTask.ts [hunk 1]
```diff
@@ -22,7 +22,7 @@ export default class AddSiteAreaLimitToConsumptionsTask extends MigrationTask {
   async migrateTenant(tenant: Tenant) {
     let modifiedCount = 0;
     // Get Charging Stations
-    const siteAreas = await SiteAreaStorage.getSiteAreas(tenant.id, { withChargingStations: true }, Constants.DB_PARAMS_MAX_LIMIT);
+    const siteAreas = await SiteAreaStorage.getSiteAreas(tenant.id, { withChargingStationsAttribute: true }, Constants.DB_PARAMS_MAX_LIMIT);
```
Comment (LucasBrazi06) [inline src/migration/tasks/AddSiteAreaLimitToConsumptionsTask.ts:25]: What is the purpose of renaming this param to withChargingStationsAttribute?

File: src/server/ocpi/ocpi-services-impl/ocpi-2.1.1/OCPIMapping.ts [hunk 2]
```diff
@@ -138,13 +138,16 @@ export default class OCPIMapping {
...
         withChargingStations: true,
+        withChargingStationsAttribute: true,
         siteIDs: [site.id],
         issuer: true
       },
       Constants.DB_PARAMS_MAX_LIMIT);
-    for (const siteArea of siteAreas.result) {
-      // Get charging stations from SiteArea
-      evses.push(...OCPIMapping.getEvsesFromSiteaArea(tenant, siteArea, options));
+    if (siteAreas.count > 0) {
```
Comment (LucasBrazi06) [inline src/server/ocpi/ocpi-services-impl/ocpi-2.1.1/OCPIMapping.ts:146]: Not needed, the result should always be an empty array.

File: src/server/ocpi/ocpi-services-impl/ocpi-2.1.1/OCPIMapping.ts [hunk 3]
```diff
@@ -159,9 +162,11 @@ export default class OCPIMapping {
     const ocpiLocationsResult: DataResult<OCPILocation> = { count: 0, result: [] };
     // Get all sites
     const sites = await SiteStorage.getSites(tenant.id, { issuer: true, withChargingStations: true }, { limit, skip });
-    // Convert Sites to Locations
-    for (const site of sites.result) {
-      ocpiLocationsResult.result.push(await OCPIMapping.convertSite2Location(tenant, site, options));
+    if (sites.count > 0) {
```
Comment (LucasBrazi06) [inline src/server/ocpi/ocpi-services-impl/ocpi-2.1.1/OCPIMapping.ts:165]: Not needed

File: src/server/ocpi/ocpi-services-impl/ocpi-2.1.1/OCPIMapping.ts [hunk 4]
```diff
@@ -181,20 +186,22 @@ export default class OCPIMapping {
       limit,
       skip
     });
+    if (tags.count > 0) {
```
Comment (LucasBrazi06) [inline src/server/ocpi/ocpi-services-impl/ocpi-2.1.1/OCPIMapping.ts:189]: Not needed

File: src/server/ocpi/ocpi-services-impl/ocpi-2.1.1/OCPIMapping.ts [hunk 5]
```diff
@@ -214,8 +221,10 @@ export default class OCPIMapping {
       limit,
       skip
     });
-    for (const transaction of transactions.result) {
-      sessions.push(transaction.ocpiData.session);
+    if (transactions.count > 0) {
```
Comment (LucasBrazi06) [inline src/server/ocpi/ocpi-services-impl/ocpi-2.1.1/OCPIMapping.ts:224]: Not needed

File: src/server/ocpi/ocpi-services-impl/ocpi-2.1.1/OCPIMapping.ts [hunk 6]
```diff
@@ -235,9 +244,11 @@ export default class OCPIMapping {
       limit,
       skip
     });
-    for (const transaction of transactions.result) {
-      if (transaction.ocpiData && transaction.ocpiData.cdr) {
-        cdrs.push(transaction.ocpiData.cdr);
+    if (transactions.count > 0) {
```
Comment (LucasBrazi06) [inline src/server/ocpi/ocpi-services-impl/ocpi-2.1.1/OCPIMapping.ts:247]: Not needed

File: src/storage/mongodb/SiteAreaStorage.ts [hunk 7]
```diff
@@ -144,7 +144,7 @@ export default class SiteAreaStorage {
   public static async getSiteAreas(tenantID: string,
     params: {
       siteAreaID?: string; search?: string; siteIDs?: string[]; withSite?: boolean; issuer?: boolean;
-      withChargingStations?: boolean; smartCharging?: boolean; withAvailableChargingStations?: boolean;
+      withChargingStationsAttribute?: boolean; withChargingStations?: boolean; withAvailableChargingStations?: boolean; smartCharging?: boolean;
```
Comment (LucasBrazi06) [inline src/storage/mongodb/SiteAreaStorage.ts:147]: Confusing when to use withChargingStations or withChargingStationsAttribute?

File: src/migration/tasks/AddSiteAreaLimitToConsumptionsTask.ts (hunk 1 above)
Comment (jerome-benoit) [reply src/migration/tasks/AddSiteAreaLimitToConsumptionsTask.ts:25]: + withChargingStationsAttribute add an attribute to the returned object containing the charging stations in the site area;
+ withChargingStations will only return site areas with charging stations attached to them. It follows the same semantic as getSites().

Comment (LucasBrazi06) [reply src/migration/tasks/AddSiteAreaLimitToConsumptionsTask.ts:25]: > * withChargingStations

withChargingStations means: get the charging stations of the site areas and return it as a property (join).
If you want to get the Site Areas which have at least one charging station, maybe use something like this withOnlyChargingStations and provide the MongoDB request

Comment (jerome-benoit) [reply src/migration/tasks/AddSiteAreaLimitToConsumptionsTask.ts:25]: > > * withChargingStations
> 
> withChargingStations means: get the charging stations of the site areas and return it as a property (join).

That param that is used in getSites(), HttpSiteAreaRequest and now getSiteAreas() does not mean at all that in the code. I have just aligned to the current semantic in getSiteAreas() after the param rename you've done in the previous related PR on getSites(). 
Changing the current semantic is fine but that's not the purpose of that PR and there's really a lot more to do to align params semantic between the REST API and the MongoDB methods. It's another topic, a new issue. Just like the use of queries only in MongoDB methods.

**Pull request merged**